src/spell/modules/spell/spell/src/mailbox.rs

```diff
@@ -0,0 +1,220 @@
+use marine_rs_sdk::marine;
+
+use fluence_spell_dtos::value::{StringValue, UnitValue};
+
+use crate::auth::is_by_spell;
+use crate::kv::primitive::read_string;
+use crate::misc::fetch_rows;
+use crate::schema::db;
+
+#[marine]
+/// `messages` contains up to `DEFAULT_MAX_MAILBOX` latest messages,
+/// sorted in the order they were pushed
+pub struct GetMailboxResult {
+    pub messages: Vec<String>,
+    pub success: bool,
+    pub error: String,
+}
+
+impl From<eyre::Result<Vec<String>>> for GetMailboxResult {
+    fn from(result: eyre::Result<Vec<String>>) -> Self {
+        match result {
+            Ok(messages) => GetMailboxResult {
+                messages,
+                success: true,
+                error: "".to_string(),
+            },
+            Err(e) => GetMailboxResult {
+                success: false,
+                error: format!("get_mailbox error: {}", e),
+                messages: vec![],
+            },
+        }
+    }
+}
+
+#[marine]
+/// Push a message to the mailbox. Mailbox keeps `DEFAULT_MAX_MAILBOX` latest messages.
+pub fn push_mailbox(message: String) -> UnitValue {
+    let call_parameters = marine_rs_sdk::get_call_parameters();
+
+    // We want to prevent anyone except this spell to store logs to its kv
+    if !is_by_spell(&call_parameters) {
+        return UnitValue::error("push_mailbox can be called only by the associated spell script");
+    }
+
+    let result: eyre::Result<()> = try {
+        let conn = db();
+        let mut statement = conn.prepare(r#" INSERT INTO mailbox (message) VALUES (?)"#)?;
+        statement.bind(1, message.as_str())?;
+        statement.next()?;
+    };
+
+    match result {
+        Ok(_) => UnitValue::ok(),
+        Err(e) => UnitValue::error(format!("Error storing mailbox message: {}", e)),
+    }
+}
+
+#[marine]
+/// Get all messages from the mailbox ordered by timestamp ascending.
+pub fn get_mailbox() -> GetMailboxResult {
+    let result: eyre::Result<Vec<String>> = try {
+        let conn = db();
+        let statement =
+            conn.prepare(r#"SELECT message FROM mailbox ORDER BY timestamp ASC, id ASC"#)?;
+        let messages: Vec<String> = fetch_rows(statement, |statement| {
+            Ok(Some(statement.read::<String>(0)?))
+        });
+
+        messages
+    };
+
+    result.into()
+}
+
+#[marine]
+/// Get the latest mailbox message and remove it from the mailbox.
+/// result.absent is true if there are no messages in the mailbox.
+pub fn pop_mailbox() -> StringValue {
```

Comment:
And this one should be read only from the spell, right?

Reply:
yeah